add daos

JimYam opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/1050

semuelle commented:
Thanks for the delivery, @JimYam. We will look into it as soon as possible. We have a bit of a backlog currently, so it might take a couple of days. 

In the meantime, could you remove the `.idea` folder from this PR?

JimYam commented:
> Thanks for the delivery, @JimYam. We will look into it as soon as possible. We have a bit of a backlog currently, so it might take a couple of days.
> 
> In the meantime, could you remove the `.idea` folder from this PR?

okay， thx.

dsm-w3f commented:
@JimYam thank you for the milestone submission and my apologies for the late response. Please see [the evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/daos_1_dsm-w3f.md) and provide proper answers and fixes. After that, let me know to continue with the evaluation.

JimYam commented:
> @JimYam thank you for the milestone submission and my apologies for the late response. Please see [the evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/daos_1_dsm-w3f.md) and provide proper answers and fixes. After that, let me know to continue with the evaluation.

thank you for your reply. I will expand test coverage soon and fix issues in doc and clippy. Because I wanted to use different licenses on different modules before, so there are two licenses. It seems that it is really unnecessary now. I will delete one of them.

JimYam commented:
> @JimYam thank you for the milestone submission and my apologies for the late response. Please see [the evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/daos_1_dsm-w3f.md) and provide proper answers and fixes. After that, let me know to continue with the evaluation.


What command did you use to generate Tests Coverage? The results generated on my side are quite different. I mean the results generated on my side are not accurate and need your help

dsm-w3f commented:
@JimYam I used `cargo tarpaulin --out html` on the root of the project.

JimYam commented:
> @JimYam I used `cargo tarpaulin --out html` on the root of the project.

Hi, except `We need a tutorial for performing system tests on the software. Something like this one.`, the tasks here have been completed.

JimYam commented:
> > @JimYam I used `cargo tarpaulin --out html` on the root of the project.
> 
> Hi, except `We need a tutorial for performing system tests on the software. Something like this one.`, the tasks here have been completed.

I don't think manual testing needs to be provided. As I said, this project is an abstraction layer for DAO, and business functions need to be implemented by specific projects according to their own needs. We will provide code examples to show developers how to use daos to create their DAO templates. In mileston 2, we will provide detailed documentation on how to use DAOS, which is explained when applying for a grant

JimYam commented:
> > > @JimYam I used `cargo tarpaulin --out html` on the root of the project.
> > 
> > 
> > Hi, except `We need a tutorial for performing system tests on the software. Something like this one.`, the tasks here have been completed.
> 
> I don't think manual testing needs to be provided. As I said, this project is an abstraction layer for DAO, and business functions need to be implemented by specific projects according to their own needs. We will provide code examples to show developers how to use daos to create their DAO templates. In mileston 2, we will provide detailed documentation on how to use DAOS, which is explained when applying for a grant https://github.com/w3f/Grants-Program/blob/master/applications/daos.md

I mean, milestone 1 focuses on how to let developers easily create their own DAO templates based on daos. This part does not require manual testing. More detailed usage documentation (for users), we will provide in milestone 2. Milestone 2 will be completed soon

dsm-w3f commented:
@JimYam thank you for the improvements. I updated the evaluation document with the current status. I still need to check the system's behavior. I saw that you provided some examples that use the code delivered. How can I run those examples to check the system's behavior? More details are in the [evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/daos_1_dsm-w3f.md).

JimYam commented:
> @JimYam thank you for the improvements. I updated the evaluation document with the current status. I still need to check the system's behavior. I saw that you provided some examples that use the code delivered. How can I run those examples to check the system's behavior? More details are in the [evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/daos_1_dsm-w3f.md).

Thank you for your reply. I will provide a manual test using daos on kico as soon as possible. But this is definitely not possible to do very detailed, just approximate.

JimYam commented:
> > @JimYam thank you for the improvements. I updated the evaluation document with the current status. I still need to check the system's behavior. I saw that you provided some examples that use the code delivered. How can I run those examples to check the system's behavior? More details are in the [evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/daos_1_dsm-w3f.md).
> 
> Thank you for your reply. I will provide a manual test using daos on kico as soon as possible. But this is definitely not possible to do very detailed, just approximate.



> @JimYam thank you for the improvements. I updated the evaluation document with the current status. I still need to check the system's behavior. I saw that you provided some examples that use the code delivered. How can I run those examples to check the system's behavior? More details are in the [evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/daos_1_dsm-w3f.md).

hey, boy. Kico project have added [manual tests](https://github.com/DICO-TEAM/dico-chain/blob/main/docs/create-vc-dao.md) for 

> @JimYam thank you for the improvements. I updated the evaluation document with the current status. I still need to check the system's behavior. I saw that you provided some examples that use the code delivered. How can I run those examples to check the system's behavior? More details are in the [evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/daos_1_dsm-w3f.md).

hey boy. Kico project have added [manual tests](https://github.com/DICO-TEAM/dico-chain/blob/main/docs/create-vc-dao.md) for VC DAO.

dsm-w3f commented:
@JimYam I'm trying to run the DICO system to perform the tests using [these steps](https://github.com/DICO-TEAM/dico-chain#development), however when I connect using the polkadot.js the extrinsic from the DICO pallets are not loaded. Any hint or step that may be missing in the tutorial for running the system? 

I could run kico alone using its executable, but it seems to be pointing to mainnet.

zxshask commented:
> @JimYam I'm trying to run the DICO system to perform the tests using [these steps](https://github.com/DICO-TEAM/dico-chain#development), however when I connect using the polkadot.js the extrinsic from the DICO pallets are not loaded. Any hint or step that may be missing in the tutorial for running the system?
> 
> I could run kico alone using its executable, but it seems to be pointing to mainnet.

I usually execute the `polkadot-launch config.json` command locally, and polkadot.js is also opened locally and executes the transaction. No existing errors were found.

zxshask commented:
> > @JimYam I'm trying to run the DICO system to perform the tests using [these steps](https://github.com/DICO-TEAM/dico-chain#development), however when I connect using the polkadot.js the extrinsic from the DICO pallets are not loaded. Any hint or step that may be missing in the tutorial for running the system?
> > I could run kico alone using its executable, but it seems to be pointing to mainnet.
> 
> I usually execute the `polkadot-launch config.json` command locally, and polkadot.js is also opened locally and executes the transaction. No existing errors were found.

The ws port of the relay chain is 9944 and 9945, and the ws port of tico is 9946

dsm-w3f commented:
@zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?

zxshask commented:
> @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?



> @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?

daos is just the project name, not a transaction. The daos project is an abstraction layer of governance, which can be used by any group, but the transactions contained in the specific group are DAO
determined by the template. In the kico project, it is a DAO about venture capital created based on multiple assets. According to the design of the template, the tokens module is needed here. You can create your own DAO template that does other things without the DAO tokens module at all.

zxshask commented:
> > @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?
> 
> > @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?
> 
> daos is just the project name, not a transaction. The daos project is an abstraction layer of governance, which can be used by any group, but the transactions contained in the specific group are DAO determined by the template. In the kico project, it is a DAO about venture capital created based on multiple assets. According to the design of the template, the tokens module is needed here. You can create your own DAO template that does other things without the DAO tokens module at all.

This is a DAO framework, it can create DAOs for any group, you can't test its functionality without a group. The transfer in the manual test is an example of daos executing external transactions. Everything in your DAO template that needs to be voted on in the DAO can be implemented in this way. In your DAO template, all you have to do is provide users with more functions

zxshask commented:
> @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?



> @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?

As a developer, you can create your DAO template through daos, you can create CultDAo, you can also create SOS. For example, the kico team created its own VC DAO, which is more powerful and scalable than CultDAO. Here, if you want to test, you need to understand how daos executes an external transaction, here just take transfer as an example.

zxshask commented:
The VC DAO in the kico network is currently the only example using daos, so I'm giving you this example. You can create your VC DAO according to the steps without testing the specific details of creation, because each DAO template will have different requirements. You assume that you have created a dao, now you need to set the Origin for each external function, and then the daos performs a series of voting-related operations through the internal function to obtain the Origin and execute the external function

zxshask commented:
All transactions that are allowed to be executed in the DAO template are external transactions of daos. For example, the VC DAO template mainly allows DAO to do these [external transactions](https://github.com/DICO-TEAM/dico-chain/blob/main/runtime/tico/src/vc.rs#L30)

zxshask commented:
> @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?



> @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?

Hey @dsm-w3f ,  zxshask  is our professional test engineer. He has done the manual test. As he said, daos is an abstract DAO framework, and manual testing should be based on the specific DAO implementation, since many details of functionality need to be refined by the DAO template developer. Manual testing is only an auxiliary function, and the VC DAO implemented by the kico team based on daos has completed all manual testing and will soon be on the main network

JimYam commented:
> @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?

Hey @dsm-w3f , zxshask is our professional test engineer. He has done the manual test. As he said, daos is an abstract DAO framework, and manual testing should be based on the specific DAO implementation, since many details of functionality need to be refined by the DAO template developer. Manual testing is only an auxiliary function, and the VC DAO implemented by the kico team based on daos has completed all manual testing and will soon be on the main network

JimYam commented:
> > @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?
> 
> Hey @dsm-w3f , zxshask is our professional test engineer. He has done the manual test. As he said, daos is an abstract DAO framework, and manual testing should be based on the specific DAO implementation, since many details of functionality need to be refined by the DAO template developer. Manual testing is only an auxiliary function, and the VC DAO implemented by the kico team based on daos has completed all manual testing and will soon be on the main network



> > @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?
> 
> Hey @dsm-w3f , zxshask is our professional test engineer. He has done the manual test. As he said, daos is an abstract DAO framework, and manual testing should be based on the specific DAO implementation, since many details of functionality need to be refined by the DAO template developer. Manual testing is only an auxiliary function, and the VC DAO implemented by the kico team based on daos has completed all manual testing and will soon be on the main network

For example, as a general idea, we would set up a referendum that requires users to lock up a certain token when they vote. However, we are not going to do the implementation here, we think it should be left to the DAO template developer to decide, in other words, he can decide to lock or consume anything in his template. This allows daos to be extensible enough to meet the requirements of different DAOs.

JimYam commented:
> > > @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?
> > 
> > 
> > Hey @dsm-w3f , zxshask is our professional test engineer. He has done the manual test. As he said, daos is an abstract DAO framework, and manual testing should be based on the specific DAO implementation, since many details of functionality need to be refined by the DAO template developer. Manual testing is only an auxiliary function, and the VC DAO implemented by the kico team based on daos has completed all manual testing and will soon be on the main network
> 
> > > @zxshask and @JimYam thank you for the answer. Using the port 9946 I could access the chain and see the pallets from the tutorial. However, I noticed that there are different functions in the example guide and in the version delivered. For example, in the createDao pallet there are only two functions available (create_dao and dao_remark) and in the example, there is at least one more function (daos). Also, the tokens pallet is not included in the source code and it is used in the tutorial. These differences between the version of the software and the tutorial are preventing me to continue with the tests. Could you please inform me which version was used to create the tutorial or upgrade the tutorial for using the exact version of the delivery?
> > 
> > 
> > Hey @dsm-w3f , zxshask is our professional test engineer. He has done the manual test. As he said, daos is an abstract DAO framework, and manual testing should be based on the specific DAO implementation, since many details of functionality need to be refined by the DAO template developer. Manual testing is only an auxiliary function, and the VC DAO implemented by the kico team based on daos has completed all manual testing and will soon be on the main network
> 
> For example, as a general idea, we would set up a referendum that requires users to lock up a certain token when they vote. However, we are not going to do the implementation here, we think it should be left to the DAO template developer to decide, in other words, he can decide to lock or consume anything in his template. This allows daos to be extensible enough to meet the requirements of different DAOs.

In the kico network, you can create your VC DAO for any NFT class or FT, but you must be the asset creator. As a developer, he can set his template to create DAOs based on anything else. You can't do manual testing without basing it on a specific DAO template, because you have to have a DAO in the first place. daos is designed to make it easy for people to create a variety of DAOs, rather than being limited to a specific DAO

dsm-w3f commented:
@JimYam and @zxshask thank you for the explanations. I was able to run all the tests now. Also the benchmarks worked well. The milestone is approved. Great job!

JimYam commented:
> @JimYam and @zxshask thank you for the explanations. I was able to run all the tests now. Also the benchmarks worked well. The milestone is approved. Great job!

Thank you very much for your recognition of daos, and we will continue to develop as promised, with interesting features coming up next.

JimYam commented:
> @JimYam and @zxshask thank you for the explanations. I was able to run all the tests now. Also the benchmarks worked well. The milestone is approved. Great job!

When can the reward of this milestone be allocated? I have already sent you invioce three weeks ago.

dsm-w3f commented:
@JimYam and @zxshask I already forwarded your invoice internally, the payment is expected to take place within two weeks but it can happen before that.

JimYam commented:
> > Hi @JimYam, I noticed that the on the invoice is missing our billing address and your billing address. Our billing address is:
> > Baarerstrasse 14, 6300, Zug.
> > Could you please add it and resend the invoice?
> > Many thanks, Federica Dubbini
> 
> I have filled in the billing Address according to your requirements, please let us know when you receive the invoice, if it is convenient. and i want to know the specific date of grant reward payment.

Many thanks.

zxshask commented:
> > > Hi @JimYam, I noticed that the on the invoice is missing our billing address and your billing address. Our billing address is:
> > > Baarerstrasse 14, 6300, Zug.
> > > Could you please add it and resend the invoice?
> > > Many thanks, Federica Dubbini
> > 
> > 
> > I have filled in the billing Address according to your requirements, please let us know when you receive the invoice, if it is convenient. and i want to know the specific date of grant reward payment.
> 
> Many thanks.

Hi, @dsm-w3f  and @fededubbi , Is there any problem with the invoice?

fededubbi commented:
Hi @zxshask,
No, the invoice is correct. It will be paid with the next payment run rather on Friday or next Monday.
Many thanks,
Regards,

Federica Dubbini
